File: src/main/java/com/lmp/shared/pricing/MoneyUtils.java
package com.lmp.shared.pricing;

import java.math.BigDecimal;
import java.math.RoundingMode;

/**
 * Constantes et utilitaires de précision monétaire centralisés.
 * <p>
 * Aligné sur la précision external ERP (Currency precision = 2, HALF_UP).
 * Toutes les opérations monétaires de LMP doivent utiliser ces constantes
 * pour garantir des résultats identiques entre LMP et l'ERP.
 *
 * <h3>Précisions par type</h3>
 * <table>
 *   <tr><td>Montants monétaires</td><td>2 décimales, HALF_UP</td></tr>
 *   <tr><td>Pourcentages</td><td>2 décimales, DOWN (pour répartition)</td></tr>
 *   <tr><td>Taux de TVA</td><td>4 décimales (stockage)</td></tr>
 *   <tr><td>Taux de change</td><td>6 décimales</td></tr>
 * </table>
 */
public final class MoneyUtils {

    // ── Précisions ──────────────────────────────────────────────────────────

    /** Précision des montants monétaires (EUR, USD, etc.) — external ERP Currency field. */
    public static final int CURRENCY_SCALE = 2;

    /** Précision des pourcentages (invoice_portion, etc.) — external ERP Percent field. */
    public static final int PERCENT_SCALE = 2;

    /** Précision des taux de TVA stockés (ex: 0.2100). */
    public static final int VAT_RATE_SCALE = 4;

    /** Précision des taux de change. */
    public static final int FX_RATE_SCALE = 6;

    // ── Mode d'arrondi par défaut ───────────────────────────────────────────

    /** Arrondi standard pour les montants monétaires — identique à flt() dans external ERP. */
    public static final RoundingMode CURRENCY_ROUNDING = RoundingMode.HALF_UP;

    /** Arrondi pour les répartitions de pourcentages (N-1 premières parts). */
    public static final RoundingMode PERCENT_ROUNDING = RoundingMode.DOWN;

    // ── Constantes ──────────────────────────────────────────────────────────

    public static final BigDecimal HUNDRED = new BigDecimal("100");
    public static final BigDecimal ZERO = BigDecimal.ZERO.setScale(CURRENCY_SCALE);

    private MoneyUtils() {}

    // ── Opérations monétaires centralisées ───────────────────────────────────

    /**
     * Arrondit un montant à la précision monétaire standard (2 décimales, HALF_UP).
     * Équivalent de {@code flt(value, precision("payment_amount"))} dans external ERP.
     */
    public static BigDecimal round(BigDecimal amount) {
        if (amount == null) return ZERO;
        return amount.setScale(CURRENCY_SCALE, CURRENCY_ROUNDING);
    }

    /**
     * Multiplie et arrondit au centime.
     * Ex: TVA, conversion, pourcentage.
     */
    public static BigDecimal multiply(BigDecimal a, BigDecimal b) {
        if (a == null || b == null) return ZERO;
        return a.multiply(b).setScale(CURRENCY_SCALE, CURRENCY_ROUNDING);
    }

    /**
     * Divise et arrondit au centime.
     * Ex: extraction HT depuis TTC, répartition.
     */
    public static BigDecimal divide(BigDecimal numerator, BigDecimal denominator) {
        if (numerator == null || denominator == null || denominator.compareTo(BigDecimal.ZERO) == 0) {
            return ZERO;
        }
        return numerator.divide(denominator, CURRENCY_SCALE, CURRENCY_ROUNDING);
    }

    /**
     * Calcule un pourcentage d'un montant — formule external ERP exacte.
     * {@code flt(grandTotal * flt(portion) / 100, precision("payment_amount"))}
     *
     * @param amount  montant de base (ex: grand_total)
     * @param percent pourcentage (ex: 33.33)
     * @return montant arrondi
     */
    public static BigDecimal percentOf(BigDecimal amount, BigDecimal percent) {
        if (amount == null || percent == null) return ZERO;
        return amount.multiply(percent)
                .divide(HUNDRED, CURRENCY_SCALE, CURRENCY_ROUNDING);
    }

    /**
     * Applique un taux multiplicatif sur un montant.
     * Ex: {@code applyRate(100, 0.21) = 21.00} (montant TVA)
     *     {@code applyRate(100, 1.21) = 121.00} (montant TTC)
     */
    public static BigDecimal applyRate(BigDecimal amount, BigDecimal rate) {
        return multiply(amount, rate);
    }

    /**
     * Extrait le montant HT d'un montant TTC.
     * {@code ttc / (1 + vatRate)}
     */
    public static BigDecimal extractHt(BigDecimal amountTtc, BigDecimal vatRate) {
        if (amountTtc == null) return ZERO;
        BigDecimal divisor = BigDecimal.ONE.add(vatRate != null ? vatRate : BigDecimal.ZERO);
        return amountTtc.divide(divisor, CURRENCY_SCALE, CURRENCY_ROUNDING);
    }

    /**
     * Calcule le montant de TVA d'un montant HT.
     * {@code ht × vatRate}
     */
    public static BigDecimal vatAmount(BigDecimal amountHt, BigDecimal vatRate) {
        if (amountHt == null || vatRate == null) return ZERO;
        return multiply(amountHt, vatRate);
    }
}

File: backend-dotnet/src/Lmp.Application/Pricing/MoneyUtils.cs
namespace Lmp.Application.Pricing;

/// <summary>
/// Centralised monetary-precision helpers. Faithful port of
/// <c>com.lmp.shared.pricing.MoneyUtils</c> — 2-decimal currency amounts,
/// HALF_UP (away-from-zero) rounding, aligned with the ERP currency precision
/// so LMP and the ERP produce identical results.
/// </summary>
public static class MoneyUtils
{
    public const int CurrencyScale = 2;

    public static readonly decimal Hundred = 100m;
    public static readonly decimal Zero = 0m;

    /// <summary>Round to standard currency precision (2 decimals, HALF_UP).</summary>
    public static decimal Round(decimal? amount)
        => amount is null ? Zero : Math.Round(amount.Value, CurrencyScale, MidpointRounding.AwayFromZero);

    /// <summary>Multiply then round to the cent.</summary>
    public static decimal Multiply(decimal? a, decimal? b)
        => a is null || b is null ? Zero : Math.Round(a.Value * b.Value, CurrencyScale, MidpointRounding.AwayFromZero);

    /// <summary>Divide then round to the cent. Returns zero on a null/zero denominator.</summary>
    public static decimal Divide(decimal? numerator, decimal? denominator)
    {
        if (numerator is null || denominator is null || denominator.Value == 0m)
        {
            return Zero;
        }

        return Math.Round(numerator.Value / denominator.Value, CurrencyScale, MidpointRounding.AwayFromZero);
    }

    /// <summary>Percentage of an amount — exact ERP formula <c>amount * percent / 100</c>.</summary>
    public static decimal PercentOf(decimal? amount, decimal? percent)
        => amount is null || percent is null
            ? Zero
            : Math.Round(amount.Value * percent.Value / Hundred, CurrencyScale, MidpointRounding.AwayFromZero);

    /// <summary>Apply a multiplicative rate to an amount.</summary>
    public static decimal ApplyRate(decimal? amount, decimal? rate) => Multiply(amount, rate);

    /// <summary>Extract the net (HT) amount from a gross (TTC) amount: <c>ttc / (1 + vatRate)</c>.</summary>
    public static decimal ExtractHt(decimal? amountTtc, decimal? vatRate)
    {
        if (amountTtc is null)
        {
            return Zero;
        }

        var divisor = 1m + (vatRate ?? 0m);
        return Math.Round(amountTtc.Value / divisor, CurrencyScale, MidpointRounding.AwayFromZero);
    }

    /// <summary>VAT amount of a net (HT) amount: <c>ht * vatRate</c>.</summary>
    public static decimal VatAmount(decimal? amountHt, decimal? vatRate)
        => amountHt is null || vatRate is null ? Zero : Multiply(amountHt, vatRate);
}
